### PR TITLE
gmx parser (_extract_dataframe) using pd.read_csv for performance

### PR DESCRIPTION
**Pull request by lee212**
pd.read_csv() shows the same behavior building pandas Dataframe with better performance. Parsing header lines remain the same but with `bytes` reading, instead of `text` mode.

### CODE REVIEW

**Comment by lee212** [comment]
Related to #82

**Comment by orbeckst** [comment]
@lee212 interesting idea. Make the tests pass and then ping me with @orbeckst so that I can have a closer look.

**Comment by lee212** [comment]
@orbeckst , two additional commits fixed test failures. I think it's ready for a review.

File: src/alchemlyb/parsing/gmx.py [hunk 1]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
+
+
+def _get_headers(xvg):
+    """Build python dictionary from header lines
+    
+    Build nested key and value store by reading header ('@') lines from a file.
+    Direct access to value provides reduced time complexity O(1).
+    '_raw_lines' keeps original text 
+
+    Note. 'rb' binary read for performance
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: I attempted to open a file with 'rb' mode in python 3 but removed during this PR for python 2 support. Docstring has updated.

File: src/alchemlyb/parsing/gmx.py [hunk 2]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
 
     """
     with anyopen(xvg, 'r') as f:
-        names = []
-        rows = []
+        headers = { '_raw_lines': [] }
         for line in f:
-            line = line.strip()
-            if len(line) == 0:
+            if line[0] == '@':
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: This was added along with the 'rb' attempt, now it's restored back with `startswith` for `str`.

File: src/alchemlyb/parsing/gmx.py [hunk 3]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
+    -------
+    headers: dict
 
     """
     with anyopen(xvg, 'r') as f:
-        names = []
-        rows = []
+        headers = { '_raw_lines': [] }
         for line in f:
-            line = line.strip()
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: Restored.

File: src/alchemlyb/parsing/gmx.py [hunk 4]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
-            if len(line) == 0:
+            if line[0] == '@':
+                _parse_header(line, headers)
+                headers['_raw_lines'].append(line)
+            elif line[0] == '#':
+                headers['_raw_lines'].append(line)
                 continue
+            # assuming to start a body section
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: Restored.

File: src/alchemlyb/parsing/gmx.py [hunk 5]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
+        df = pd.DataFrame(df, columns=cols)
+
+    return df
+
+
+def _parse_header(line, headers={}, depth=2):
+    """Build python dictionary for single line header in a binary format
+
+    '_val' key contains actual value to search in the dictionary.
+    Note. No return value but 'headers' dictionary will be updated
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: This was experimental to see if it reduces parsing time with python dictionary. No need to replace the old code. BTW, I added an example.

File: src/alchemlyb/parsing/gmx.py [hunk 6]
```diff
@@ -227,38 +237,121 @@ def _extract_legend(xvg):
...
+
+    """
+    if headers is None:
+        headers = _get_headers(xvg)
+    xaxis = _get_value_by_key(headers, 'xaxis', 'label')
+    names = [_get_value_by_key(headers, 's{}'.format(x), 'legend') for x in
+            range(len(headers)) if 's{}'.format(x) in headers]
+    cols = [xaxis] + names
+    header_cnt = len(headers['_raw_lines'])
+    df = pd.read_csv(xvg, sep=r"\s+", header=None, skiprows=header_cnt,
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/alchemlyb/parsing/gmx.py]: `.bz2` and `.gz` were tested, I need to confirm `.zip`. datasets from `Alchemtest` were all compressed, though.

**Comment by orbeckst** [comment]
Many thanks @lee212 ! COngrats on first merged PR!

**Pull request merged**